### PR TITLE
fix: Wrong evaluated value after binding checkbox group widget with…

### LINKED ISSUE #24620: [Bug]: Wrong evaluated value after binding checkbox group widget with query
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

When we bind a checkbox group widget with a query, the default options seem to be retained in the evaluation.

### Steps To Reproduce

1. D&D checkbox group & text widgets.
2. In the Text property for the text widget enter: `{{CheckboxGroup1.selectedValues}}`.
3. Now select all values in the checkbox group widget. All selected values should reflect in text widget.
4. Now, remove all the options & bind the checkbox query with any datasource.

**Expected:** All default options should disappear & only the options selected from the datasource should reflect in the text widget.
**Actual:** We are able to see old options as well.

### Public Sample App

_No response_

### Environment

Production

### Issue video log

https://www.loom.com/share/bc63bae4935442d98f690326d35c7a6e

### Version

Cloud
```[tasklist]
### Tasks
- [ ] https://github.com/appsmithorg/appsmith/issues/25849
```

### PR DESCRIPTION
**Pull request by Shivam-z**
## Description

- In this PR I have fixed wrong evaluated value after binding checkbox group widget with query

Output: [Loom](https://www.loom.com/share/ddcc10c5ab8b4ead8ef5d10843671d95?sid=6346b4a9-9a89-4c7d-bdcd-73c428f17626)

> [!TIP]  
> _Add a TL;DR when the description is longer than 500 words or extremely technical (helps the content, marketing, and DevRel team)._
>
> _Please also include relevant motivation and context. List any dependencies that are required for this change. Add links to Notion, Figma or any other documents that might be relevant to the PR._

Fixes #24620  
_or_  
Fixes https://github.com/appsmithorg/appsmith/issues/24620
> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags=""

### :mag: Cypress test results

> [!CAUTION]  
> If you modify the content in this section, you are likely to disrupt the CI result for your PR.

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [x] No

## Summary by CodeRabbit

- **New Features**
	- Introduced comprehensive snapshot tests for Anvil Widgets in Canvas, Preview, and Deploy modes.
	- Added test cases for Anvil modals in the Anvil editor.
	- Enabled filtering functionality for TableV2 widgets by default.

- **Bug Fixes**
	- Addressed issues with Git sync tests by updating selectors and function calls.
	- Fixed control flow in various Cypress tests by removing `cy.pause()`.

- **Chores**
	- Updated Dockerfile to ensure custom command-scripts have executable permissions.
	- Added environment variable `DP_POSTGRES_URL` to Docker deployment configuration.

- **Refactor**
	- Refined job configurations and script executions in GitHub Actions workflows for improved performance and clarity.

### CODE REVIEW

File: app/client/src/widgets/CheckboxGroupWidget/widget/index.tsx [hunk 1]
```diff
@@ -32,6 +32,7 @@ import IconSVG from "../icon.svg";
 import ThumbnailSVG from "../thumbnail.svg";
 import { WIDGET_TAGS } from "constants/WidgetConstants";
 import { FlexVerticalAlignment } from "layoutSystems/common/utils/constants";
+import _ from "lodash";
```
[the remark this answers is not in the document]
Comment (Shivam-z) [reply app/client/src/widgets/CheckboxGroupWidget/widget/index.tsx:35]: done

File: app/client/src/widgets/CheckboxGroupWidget/widget/index.tsx [hunk 2]
```diff
@@ -9,7 +9,7 @@ import type { ValidationResponse } from "constants/WidgetValidation";
 import { ValidationTypes } from "constants/WidgetValidation";
 import type { SetterConfig, Stylesheet } from "entities/AppTheming";
 import { EvaluationSubstitutionType } from "entities/DataTree/dataTreeFactory";
-import { compact, xor } from "lodash";
+import { compact, xor, findIndex } from "lodash";
 import { default as React } from "react";
```
[the remark this answers is not in the document]
Comment (Shivam-z) [reply app/client/src/widgets/CheckboxGroupWidget/widget/index.tsx:13]: The imports from lodash are not  used as types , plese keep it as it is.

File: app/client/src/widgets/CheckboxGroupWidget/widget/index.tsx [hunk 3]
```diff
@@ -614,6 +614,26 @@ class CheckboxGroupWidget extends BaseWidget<
...
+          findIndex(
+            prevProps.options,
+            (option: OptionProps) => option.value === i,
+          ) === -1
+        ) {
+          this.props.updateWidgetMetaProperty("selectedValues", []);
+          break;
+        }
+      }
+    }
```
[the remark this answers is not in the document]
Comment (Shivam-z) [reply app/client/src/widgets/CheckboxGroupWidget/widget/index.tsx:635]: done

File: app/client/src/widgets/CheckboxGroupWidget/widget/index.tsx [hunk 4]
```diff
@@ -614,6 +614,18 @@ class CheckboxGroupWidget extends BaseWidget<
...
+      (value: string) =>
+        prevProps.options.some((option) => option.value === value),
+    );
+    if (validSelectedValues.length !== prevProps.selectedValues.length) {
+      this.props.updateWidgetMetaProperty(
+        "selectedValues",
+        validSelectedValues,
+      );
+    }
+
```
[the remark this answers is not in the document]
Comment (Shivam-z) [reply app/client/src/widgets/CheckboxGroupWidget/widget/index.tsx:628]: not required

**Comment by ramsaptami** [comment]
Fix works on fixed mode, needs to be propagated to Anvil too 
cc: @jsartisan @KelvinOm

**Comment by jsartisan** [comment]
@Shivam-z The solution does not work properly. Check the video. A better approach would be a custom validator for the defaultSelectedValues property.

https://github.com/appsmithorg/appsmith/assets/6636360/4962a637-8919-408f-9437-0ed1cec838ef

**Comment by jsartisan** [comment]
Closing the PR as the solution does not solve it properly.

**Comment by Shivam-z** [comment]
@jsartisan sorry, I was bit busy with the other work , i.e could not able to looked into this PR.
can you give me some time to look into this issue and for a better solution.
Please reopen this PR.

**Comment by Shivam-z** [comment]
@jsartisan I have modified defaultSelectedValuesValidation
could you please review that.

**Comment by jsartisan** [comment]
The solution looks good to me. Two things
1. the validation is always valid. even when the value is not an array.
2. When the options change, it does not trigger the validation again.
The solution to this is to add depencyMap:
```
static getDependencyMap(): Record<string, string[]> {
    return {
      defaultSelectedValues: ["options"],
    };
  }
 ```

**Comment by Shivam-z** [comment]
Hi @jsartisan , I fixed the  second suggestion you gave me.
can you give me more clarification on first suggestion "the validation is always valid. even when the value is not an array."

here which value you are talking about ? is it defaultSelectedValues ?
If it is **defaultSelectedValues** , then in case of checkBoxGroup Widget it should be an array  only right ?

**Comment by jsartisan** [comment]
Hey, couldn't get time to check again. Regarding  this"the validation is always valid. even when the value is not an array.", i think it's fine.

@ramsaptami can you check this PR as well? Codewise it looks okay. Need some help in checking validations for queries.

**Comment by ramsaptami** [comment]
- default value when different from derived value is not retained in `selectedValues` property
- binding a different query refreshes the 'selectedValues' property
image

cc: @jsartisan

**Comment by Shivam-z** [comment]
@jsartisan can you please look into this PR and tell me whether it is good or I need to rectify something.

**Comment by jsartisan** [comment]
@Shivam-z Can you check the comments by Saptami? The solution is failing in a few cases.

**Comment by Shivam-z** [comment]
@jsartisan The comments given by @ramsaptami is the expected behaviour not the bug.
 - In the previous version it was showing the defaultValues as selectedValues even though the derived did not contain those default values . For ex: derivedValues: {1,2,3} , defaultValues:{2,3,4} . so selectedValues should be {2,3} but previously it was {2,3,4} i.e 4 was not retained.
 - In the previous version when binding a different query it was not refreshing the 'selectedValues' property but now it is fixed.

**Review by jsartisan — DISMISSED** [review]

**Comment by jsartisan** [comment]
@Shivam-z Code looks good to me now. I ran the tests on a separate PR and found there is one test that is constantly failing

`cypress/e2e/Regression/ClientSide/Widgets/Switch/SwitchGroup1_spec.ts`

https://github.com/appsmithorg/appsmith/assets/6636360/c9d1efbd-a662-42fb-966c-7823d2cb992c

Check the video. 

Btw Sorry for the back and forth, I know it's taking time.

**Comment by Shivam-z** [comment]
Hi @jsartisan , I have updated some code in  the cypress test case for **SwitchGroup1_spec.ts** file and it is running  fine in my local.  can you please check this PR.

**Comment by jsartisan** [comment]
@Shivam-z something wrong with your PR. It's now showing 616 files. Can you rebase latest release into your branch

**Comment by jsartisan** [comment]
I want to create a PR from your PR and run tests on it like we did previously but now your PR is showing 616 files changed.

**Comment by Shivam-z** [comment]
@jsartisan updated file changes

**Review by jsartisan — DISMISSED** [review]

**Comment by jsartisan** [comment]
@Shivam-z all tests passing now. What was the issue?

**Comment by Shivam-z** [comment]
Hi @jsartisan , some assertions were failing in `Switch/SwitchGroup1_spec.ts` and  it was expecting different color value because of the functionality added in this PR and due to this the test cases were failing earlier , but now it is fixed.
Regarding test to cover this functionality , I have already added one test case in : `app/client/src/widgets/CheckboxGroupWidget/widget/index.test.tsx`

**Pull request merged**